sys: Add Disjoint-set (union-find) data structure

rruuaanng opened the pull request:
Add a set of Disjoint-set functions (`find`, `union`) to handle queries between sets.

henrikbrixandersen reviewed (commented):
New features should have test cases added.

kartben commented on include/zephyr/sys/uf.h:
```diff
@@ -0,0 +1,60 @@
+	/** @endcond */
+};
+
+/**
+ * @brief Initialize a disjoint-set.
+ */
+static inline void uf_makeset(struct uf_node *node)
+{
+	node->parent = node;
+	node->rank = 0;
+}
+
+/**
+ * @brief Find the root of the disjoint-set.
+ */
+struct uf_node *uf_find(struct uf_node *node);
+
+/**
+ * @brief Merge two nodes into the same disjoint-set.
+ */
+void uf_union(struct uf_node *node1, struct uf_node *node2);
```
Needs full Doxygen comments, including parameters and return values

andyross commented on include/zephyr/sys/uf.h:
```diff
@@ -0,0 +1,60 @@
+ */
+struct uf_node {
+	/** @cond INTERNAL_HIDDEN */
+	struct uf_node *parent;
```
This looks suspiciously like an slist, no?  Not like linked list handling is that big a deal, but every byte of code savings helps.

rruuaanng replied:
Yes, but it's dynamically sized :(

andyross reviewed (commented):
Design note on the pointer handling.

Also count me on team `sys_set_`.  "`uf`" is an expression of deep exasperation.

cfriedt reviewed (commented):
@rruuaanng - just a couple of nits about naming

cfriedt commented on tests/lib/data_structure/set/src/main.c:
```diff
@@ -0,0 +1,37 @@
+	struct sys_set_node node;
+} data_list[5] = {{.n = 1}, {.n = 3}, {.n = 2}, {.n = 5}, {.n = 4}};
+
+ZTEST(set, test_find_and_union)
+{
+	sys_set_makeset(&data_list[0].ufnode, 5);
+	sys_set_makeset(&data_list[1].ufnode, 4);
+	sys_set_makeset(&data_list[2].ufnode, 3);
+	sys_set_makeset(&data_list[3].ufnode, 2);
+	sys_set_makeset(&data_list[4].ufnode, 1);
+
+	sys_set_union(&data_list[0].ufnode, &data_list[1].ufnode);
+	sys_set_union(&data_list[1].ufnode, &data_list[2].ufnode);
+	sys_set_union(&data_list[2].ufnode, &data_list[3].ufnode);
+	sys_set_union(&data_list[3].ufnode, &data_list[4].ufnode);
+
+	struct sys_set_node *root_node = sys_set_find(&data_list[4].ufnode);
+	struct user_data *root_data = CONTAINER_OF(root_node, struct user_data, ufnode);
+
+	zassert_equal(root_data->n, data_list[0].n, "Root set data n are not equal");
+	zassert_equal(root_data->ufnode.rank, root_node->rank, "Root set are not equal");
+}
```
You will need to rename ufnode to node everywhere

cfriedt reviewed (dismissed):
@rruuaanng - please also add a testcase.yaml file.

It may be worth it to add incremental tests to each iteration in the one test function to validate certain assumptions about the disjoint set.

E.g. rank at each stage, parent properties, and so on.

jukkar reviewed (dismissed):
LGTM

nashif commented:
and move the release notes to 4.4

rruuaanng committed:
sys: Add Disjoint-set data structure

Add a set of Disjoint-set functions (`find`, `union`)
to handle queries between sets.

Signed-off-by: James Roy <rruuaanng@outlook.com>

rruuaanng committed:
tests: sys: Add unit test for Disjoint-set

Add unit tests for functions (`sys_set_makeset`,
`sys_set_union` and `sys_set_find`).

Signed-off-by: James Roy <rruuaanng@outlook.com>

andyross reviewed (dismissed):
This has been kicking around for a while.  Don't see anything I'd complained about that would prevent merge.

cfriedt reviewed (commented):
I think the name of the testsuite should be unique. Removing my earlier nak
